### PR TITLE
Uplift various `*Predicate` types into `rustc_type_ir`

### PR DESCRIPTION
**Pull request by compiler-errors**
Uplifts `ProjectionPredicate`, `ExistentialTraitRef`, `ExistentialProjection`, `TraitPredicate`, `NormalizesTo`, `CoercePredicate`, and `SubtypePredicate`.

Adds `rustc_type_ir_macros`, which semi-duplicates the derive for `TypeVisitable`, `TypeFoldable`, and `Lift`, but in a way that is interner-agnostic.

Moves `rustc_type_ir::trait_ref` to `rustc_type_ir::predicate`. The specific placement of all these structs doesn't matter b/c of glob imports, tho.

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? lcnr

File: compiler/rustc_middle/src/ty/predicate.rs [hunk 1]
```diff
@@ -1,22 +1,34 @@
...
 use rustc_type_ir::ClauseKind as IrClauseKind;
+use rustc_type_ir::CoercePredicate as IrCoercePredicate;
+use rustc_type_ir::ExistentialProjection as IrExistentialProjection;
+use rustc_type_ir::ExistentialTraitRef as IrExistentialTraitRef;
+use rustc_type_ir::NormalizesTo as IrNormalizesTo;
 use rustc_type_ir::PredicateKind as IrPredicateKind;
+use rustc_type_ir::ProjectionPredicate as IrProjectionPredicate;
+use rustc_type_ir::SubtypePredicate as IrSubtypePredicate;
+use rustc_type_ir::TraitPredicate as IrTraitPredicate;
 use rustc_type_ir::TraitRef as IrTraitRef;
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/predicate.rs:14]:
```rust
use rustc_type_ir as ir;
```
and then use `ir::ClauseKind` in the type defs 🤔

File: compiler/rustc_type_ir/src/inherent.rs [hunk 2]
```diff
@@ -50,7 +51,12 @@ pub trait GenericsOf<I: Interner<GenericsOf = Self>> {
 }
 
 pub trait GenericArgs<I: Interner<GenericArgs = Self>>:
-    Copy + DebugWithInfcx<I> + Hash + Eq + IntoIterator<Item = I::GenericArg>
+    Copy
+    + DebugWithInfcx<I>
+    + Hash
+    + Eq
+    + IntoIterator<Item = I::GenericArg>
+    + Deref<Target: Deref<Target = [I::GenericArg]>>
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/inherent.rs:59]: :/ `Deref<Target: Deref<Target = [I::GenericArg]>>` is not ideal 🤔 don't have any immediate better ideas though

**Review by lcnr — COMMENTED** [review]
r=me with or without nit applied

**Comment by lcnr** [comment]
please also update PR name

File: compiler/rustc_type_ir/src/inherent.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_type_ir/src/inherent.rs:59]: Yes it extremely sucks and also is likely not going to work for r-a, but we should fix it later (either adapt the callsites that rely on this to do something else, or something idk).

**Pull request merged**